File: src/Solutions/Euler214.hs
module Solutions.Euler214 where
import Utils.NumberTheory(totientArrayUpTo, unPrime, primesUpTo)
import Utils.Array(funArray)
import Data.Array as B
import Data.Array.Unboxed as U

bound = 40000000
totient = totientArrayUpTo bound

chainLength :: Array Int Int
chainLength = funArray 1 bound f

f :: Int -> Int
f 1 = 1
f n = 1 + chainLength B.! (totient U.! n)

euler214 :: IO String
euler214 = return $ show $ sum $ filter ((==25) . (chainLength U.!)) $ map unPrime $ primesUpTo bound

File: rust/src/solutions/euler214.rs
use crate::utils::number_theory::totient::totient_array;

pub fn euler214() -> String {
    let bound = 40000000;
    let mut chain = Vec::<u32>::with_capacity(bound);
    chain.push(1);
    for tot in totient_array::<usize>(bound).into_iter().skip(1) {
        chain.push(1 + chain[tot - 1]);
    }
    primal::Sieve::new(bound)
        .primes_from(0)
        .take_while(|&p| p < bound)
        .filter(|&p| chain[p - 1] == 25)
        .sum::<usize>()
        .to_string()
}
